edgelet/edgelet-config/src/lib.rs

```diff
@@ -128,14 +128,138 @@ impl Manual {
 #[derive(Debug, Deserialize, Serialize)]
+#[serde(tag = "method")]
 #[serde(rename_all = "lowercase")]
+pub enum AttestationMethod {
+    Tpm(TpmAttestationInfo),
+    #[serde(rename = "symmetric_key")]
+    SymmetricKey(SymmetricKeyAttestationInfo),
+    X509(X509AttestationInfo),
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+#[serde(rename_all = "lowercase")]
+pub struct TpmAttestationInfo {
+    registration_id: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    device_id: Option<String>,
+}
+
+impl TpmAttestationInfo {
+    pub fn new(registration_id: String) -> Self {
+        TpmAttestationInfo {
+            registration_id,
+            device_id: None,
+        }
+    }
+
+    pub fn registration_id(&self) -> &str {
+        &self.registration_id
+    }
+
+    pub fn device_id(&self) -> Option<&str> {
+        self.device_id.as_ref().map(AsRef::as_ref)
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+#[serde(rename_all = "lowercase")]
+pub struct SymmetricKeyAttestationInfo {
+    registration_id: String,
+    symmetric_key: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    device_id: Option<String>,
+}
+
+impl SymmetricKeyAttestationInfo {
+    pub fn registration_id(&self) -> &str {
+        &self.registration_id
+    }
+
+    pub fn symmetric_key(&self) -> &str {
+        &self.symmetric_key
+    }
+
+    pub fn device_id(&self) -> Option<&str> {
+        self.device_id.as_ref().map(AsRef::as_ref)
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+#[serde(rename_all = "lowercase")]
+pub struct X509AttestationInfo {
+    identity_cert: PathBuf,
+    identity_pk: PathBuf,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    device_id: Option<String>,
+}
+
+impl X509AttestationInfo {
+    pub fn identity_cert(&self) -> &Path {
+        &self.identity_cert
+    }
+
+    pub fn identity_pk(&self) -> &Path {
+        &self.identity_pk
+    }
+
+    pub fn device_id(&self) -> Option<&str> {
+        self.device_id.as_ref().map(AsRef::as_ref)
+    }
+}
+
+#[derive(Debug, Serialize)]
 pub struct Dps {
     #[serde(with = "url_serde")]
     global_endpoint: Url,
     scope_id: String,
-    registration_id: String,
-    #[serde(skip_serializing_if = "Option::is_none")]
-    symmetric_key: Option<String>,
+    attestation: AttestationMethod,
+}
+
+impl<'de> serde::Deserialize<'de> for Dps {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        #[derive(Debug, serde_derive::Deserialize)]
+        struct Inner {
+            #[serde(with = "url_serde")]
+            global_endpoint: Url,
+            scope_id: String,
+            registration_id: Option<String>,
+            #[serde(skip_serializing_if = "Option::is_none")]
+            attestation: Option<AttestationMethod>,
+        }
+
+        let value: Inner = serde::Deserialize::deserialize(deserializer)?;
+
+        let attestation = match value.attestation {
```

Comment:
Another more ML-ish way to write this would be:

```rust
let attestation = match (value.attestation, value.registration_id) {
    (Some(att), Some(_)) => return Err(...),
    (Some(att), None) => att,
    (None, Some(reg_id) => AttestationMethod::Tpm(...),
    (None, None) => return Err(...),
};
```

But the current one is also fine.

Reply:
This is a good suggestion. Thanks, I will update this. I makes this quite readable and pleasant :).

Reply:
Will need your re-approval please.